src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,438 @@
+        await joinerDAO.addCandidate(partId, candidate);
+        await signOut();
+
+        await createConnectedGoogleUser('malicious@bar.com', 'malicious');
+
+        // When the malicious users tries to change another candidate's fields
+        const update: Partial<MinimalUser> = { name: 'foo' };
+        const result: Promise<void> = joinerDAO.subCollectionDAO(partId, 'candidates').set(candidate.id, update);
+
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should forbid creator to add themself to the candidates', async() => {
+        // Given a part and joiner, with the current user being the creator
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+
+        // When trying to add itself to the candidates
+        const result: Promise<void> = joinerDAO.addCandidate(partId, creator);
+
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should forbid non-verified user to add themselves to the candidates', async() => {
+        // Given a part, and a non-verified user
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        const joiner: Joiner = { ...JoinerMocks.INITIAL, creator };
+        await joinerDAO.set(partId, joiner);
+        await signOut();
+
+
+        const user: FireAuth.User = await createUnverifiedUser('bar@bar.com', 'user');
+        const nonVerifiedUser: MinimalUser = { id: user.uid, name: 'user' };
+        // When adding the non-verified user to the candidates
+        const result: Promise<void> = joinerDAO.addCandidate(partId, nonVerifiedUser);
+
+        // Then it should fail
+        await expectFirebasePermissionDenied(result);
+    });
+    it('should forbid non-creator to change other fields than status', async() => {
+        // Given a part, and a user (that is not the creator)
+        const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+        const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+        const partId: string = await partDAO.create(PartMocks.INITIAL);
+        await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+        await signOut();
+
+        const candidateUser: FireAuth.User = await createConnectedGoogleUser('bar@bar.com', 'candidate');
+        const candidate: MinimalUser = { id: candidateUser.uid, name: 'candidate' };
+        // Add ourself to candidate list to try to have more privileges
```

Comment:
"to try to have more privileges" ?

Reply:
car un candidat peut potentiellement avoir plus de privilèges qu'un non-candidat !